### ISSUE TITLE
Feature Request: Application Memory Dump

### ISSUE THREAD

**Issue by aarondandy** [microsoft/azure-container-apps#617]
**Is your feature request related to a problem? Please describe**.  
Sometimes software has memory leaks and other issues that impact application stability and performance. Being able to easily get a process memory dump greatly improves the chances of finding the issue and correcting it. Here is an example of an issue we recently encountered: https://github.com/dotnet/SqlClient/issues/1810 . It was extremely difficult to get a memory dump from the container, even after getting a terminal session within it, because getting a memory dump in combination with the memory leak being an issue caused replica restarts before we could complete and extract a dump. I had to migrate the background worker off of ACA and back onto Azure Web Apps to successfully and reliably get memory dumps in production to find the issue.

**Describe the solution you'd like**.    
I would like an easy way within the Azure Portal to generate a memory dump for a process running within a container and download it to my local workstation while being able to use normal dotnet/aspnet container base images for our dotnet applications. Example: https://azure.github.io/AppService/2021/11/01/Diagnostic-Tools-for-ASP-NET-Core-Linux-apps-are-now-publicly-available.html#collection-in-kudu

**Describe alternatives you've considered**.   
We have migrated off of ACA and back to Azure Web Apps in this situation and used their `/newui` tool to download a dotnet process memory dump: https://techcommunity.microsoft.com/t5/apps-on-azure-blog/new-kudu-ui-for-app-service-on-linux-preview/ba-p/3212270

**Additional context**.  
- https://github.com/microsoft/azure-container-apps/issues/599 may be a prerequisite

**Comment by TrayanZapryanov** [comment]
Any news here?
We also have similar problem and need dump functionality to understand the problem.
Having 10 replicas working well and all of them behave similarly, the only one is constantly doubling CPU and memory.
There is something specific on it, but without several dumps it is really hard to solve the problem only with metrics and log.
Please suggest at least some workaround how to take a dump, event in 10 or more steps.

**Comment by pmsousa** [comment]
@SophCarp can you provide some insight about this? It would be valuable for Dev Teams to debug issues with container apps

**Comment by AdeZwart** [comment]
We also need this in our team! Please add something for it.

**Comment by EinmalIM** [comment]
Yes, please, give us an option to create memory dumps to analyse Container App instances

**Comment by maskati** [comment]
You can use [dotnet-dump](https://learn.microsoft.com/en-us/dotnet/core/diagnostics/dotnet-dump) to create memory dumps.
```sh
# install dotnet-dump
wget -qO /usr/local/bin/dotnet-dump https://aka.ms/dotnet-dump/linux-musl-x64 && chmod +x /usr/local/bin/dotnet-dump
# create memory dumps of all .net processes
dotnet-dump ps | awk 'NF > 0 {print $1}' | while read pid; do dotnet-dump collect -p $pid -o "/tmp/dotnet-dump-$pid.dmp";done
# analyze the dumps
ls /tmp/dotnet-dump-*.dmp | while read dumpfile; do dotnet-dump analyze $dumpfile -c dumpheap -c exit;done
```

For example running against [mcr.microsoft.com/dotnet/samples:aspnetapp](https://mcr.microsoft.com/en-us/artifact/mar/dotnet/samples/about):
```
...
7f1c57726bc8   147    70,288 System.Char[]
7f1c5b13afc0   261    71,824 System.Reflection.CustomAttributeNamedParameter[]
7f1c59387d40   781    74,976 System.Reflection.RuntimeParameterInfo
7f1c57589db8 1,999    79,960 System.RuntimeType
7f1c587d0978   429    81,367 System.Byte[]
7f1c5916c340 1,018   105,872 System.Reflection.RuntimeMethodInfo
7f1c5758bf68   604   106,824 System.Object[]
7f1c58a56e60   909   145,440 System.RuntimeType+RuntimeTypeCache
7f1c576389e0   526   230,356 System.Int32[]
7f1c5763d7c8 4,576   363,814 System.String
Total 38,848 objects, 3,132,778 bytes
```

Alternatively use [dotnet-monitor](https://github.com/dotnet/dotnet-monitor).

**Comment by electroma** [comment]
With [debug console](https://learn.microsoft.com/en-us/azure/container-apps/container-debug-console) (either from CLI or Azure Portal) the process is not too painful and works with distroless containers. Dumps may be uploaded into the BLOB container or file share for further analysis (azcopy or [blobfuse](https://learn.microsoft.com/en-us/azure/storage/blobs/blobfuse2-what-is) mount may work to write directly into the blob). A variation of @maskati's script:

```bash
# install dotnet-dump and deps, then switch to the user context to access application container filesystem
tdnf install -y azcopy dotnet-sdk-8.0 && wget https://aka.ms/dotnet-dump/linux-x64 -qO /usr/local/bin/dotnet-dump && chmod a+rx /usr/local/bin/dotnet-dump && switch-user
# create memory dumps of all .net processes
export HOME=/tmp TMPDIR=/proc/1/root/tmp && dotnet-dump ps | awk 'NF > 0 {print $1}' | while read pid; do dotnet-dump collect -p $pid -o "/tmp/dotnet-dump-$pid.dmp";done
# push the dumps into the storage account
azcopy copy /proc/1/root/tmp/  "https://<account>.blob.core.windows.net/<container>/<SAS>" --recursive --include-pattern "dotnet-dump*.dmp"
```
